### PR TITLE
Use Tor's ADD_ONION command when available

### LINKED ISSUE #227: Use the new ADD_ONION control API when available
Tor's ADD_ONION API will enable lots of cool things for us, like private key encryption, and using system-level tor instances.

It's only available in Tor 0.2.7.x (currently alpha).

### PR DESCRIPTION
**Pull request by special**
For Tor >= 0.2.7, start using the ADD_ONION control command instead of SETCONF to configure hidden services, and store the private keys in our configuration file instead of as separate files on the filesystem.

Older versions of Tor remain supported, and the behavior will not change. There is an edge case where a profile is created with a new version of Tor and later downgraded - this is not supported. Don't do that.

Existing profiles will copy their keys into the configuration and stop using the files on disk, but those files will not be removed yet to help avoid any risk of data loss. They will be removed later, likely when we enable profile encryption.

Fixes #227

### CODE REVIEW

**Comment by ioerror** [comment]
What kind of tests do you want to see here to merge this code?

File: src/tor/AddOnionCommand.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    QByteArray out("ADD_ONION");
+
+    if (m_service->privateKey().isLoaded()) {
+        out += " RSA1024:";
+        out += m_service->privateKey().encodedPrivateKey(CryptoKey::DER).toBase64();
+    } else {
+        out += " NEW:RSA1024";
+    }
+
+    foreach (const HiddenService::Target &target, m_service->targets()) {
```
Comment (rburchell) [inline src/tor/AddOnionCommand.cpp]: not something you have to change now, but in the future you may want to look into range-for (and qAsConst), it avoids copies of data & leads to smaller binary size.

File: src/tor/AddOnionCommand.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        out += " Port=";
+        out += QByteArray::number(target.servicePort);
+        out += ",";
+        out += target.targetAddress.toString().toLatin1();
```
Comment (rburchell) [inline src/tor/AddOnionCommand.cpp]: is this really supposed to be latin1?

File: src/tor/AddOnionCommand.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    out.append("\r\n");
+    return out;
+}
+
+void AddOnionCommand::onReply(int statusCode, const QByteArray &data)
+{
+    TorControlCommand::onReply(statusCode, data);
+    if (statusCode != 250) {
```
Comment (rburchell) [inline src/tor/AddOnionCommand.cpp]: would be nice to have a constant rather than a magic number

**Comment by rburchell** [comment]
:+1:

**Comment by ioerror** [comment]
Works for me on SubgraphOS now. Thanks!

**Pull request merged**